pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/security/ApexSharingViolationsRule.java

```diff
@@ -21,45 +30,103 @@
  */
 public class ApexSharingViolationsRule extends AbstractApexRule {
 
+    /**
+     * Keep track of previously reported violations to avoid duplicates.
+     */
     private WeakHashMap<ApexNode<?>, Object> localCacheOfReportedNodes = new WeakHashMap<>();
 
     public ApexSharingViolationsRule() {
         setProperty(CODECLIMATE_CATEGORIES, "Security");
         setProperty(CODECLIMATE_REMEDIATION_MULTIPLIER, 100);
         setProperty(CODECLIMATE_BLOCK_HIGHLIGHTING, false);
+        addRuleChainVisit(ASTDmlDeleteStatement.class);
+        addRuleChainVisit(ASTDmlInsertStatement.class);
+        addRuleChainVisit(ASTDmlMergeStatement.class);
+        addRuleChainVisit(ASTDmlUndeleteStatement.class);
+        addRuleChainVisit(ASTDmlUpdateStatement.class);
+        addRuleChainVisit(ASTDmlUpsertStatement.class);
+        addRuleChainVisit(ASTMethodCallExpression.class);
+        addRuleChainVisit(ASTSoqlExpression.class);
+        addRuleChainVisit(ASTSoslExpression.class);
     }
 
     @Override
-    public Object visit(ASTUserClass node, Object data) {
```

Comment:
The previous implementation used the outer classes sharing declaration and never visited the inner ASTUserClass, because it didn't use RuleChains or call `super.visit(node, data);`, 

Adding rule chains or calling super.visit was not enough because the logic for finding the child nodes doesn't distinguish between statements in the Outer or Inner class.

I tried an implementation that kept top down approach by filtering out child nodes based on the parent, but it was messy. Changing the logic to visit the Database/DML/SOQL/SOSL nodes, then look up the tree for the closes parent class makes the logic cleaner.

Reply:
This makes sense. Using node.getFirstParent() is simpler.